### PR TITLE
Preserve strides for functionalized in-place mutations

### LINKED ISSUE #133512: functionalization doesn't faithfully reproduce strides of intermediate values
This can lead to silent incorrectness for operators that are size-sensitive:
```py
import torch

@torch.compile(backend="aot_eager", fullgraph=True)
def f(x):
    x.add_(1)
    result = torch.as_strided_copy(x, (3,), (2,))
    return result

x = torch.arange(10)[::2]
y = f(x)
print(y)

```
this gives different results with or without torch.compile

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186328

Functionalization rewrites an in-place op such as `x.add_(1)` into its
functional variant followed by `replace_`. The functional variant can return a
fresh contiguous tensor even when eager in-place mutation would preserve the
original tensor's size, stride, and storage offset. If a later operation is
stride-sensitive, such as `as_strided_copy`, the functionalized program can read
from a different logical layout and silently produce different values.

Fix this by preserving the old strided layout when generated mutation handling
replaces a functional tensor with the functional op result. When the old and new
values are strided tensors with the same sizes but different stride metadata, we
copy the new values into the old layout and replace with that tensor. This keeps
the common dense-stride path unchanged while matching eager semantics for
non-contiguous mutated tensors. The TensorList overload handles generated
multi-mutation wrappers through the same helper.

Fixes #133512

Generated by my agent

Benchmark Results:
CPU microbenchmark using `torch.compile(backend="aot_eager", fullgraph=True)`,
N=200000, 200 iterations, 5 repeats, after rebuilding each revision with
`ninja -C build torch_cpu torch_python install`.

Before:
- contiguous_add_cos: median 360.23 us, runs [369.60, 358.80, 358.83, 363.21, 360.23]
- strided_add_as_strided_copy: median 380.52 us, runs [384.55, 375.85, 376.60, 380.52, 382.06]

After:
- contiguous_add_cos: median 377.17 us, runs [371.83, 352.78, 383.20, 384.06, 377.17]
- strided_add_as_strided_copy: median 717.78 us, runs [720.04, 707.56, 716.13, 717.78, 754.08]

The contiguous case is effectively unchanged within run noise. The strided case
is slower because the fix performs the necessary stride-preserving copy; the
baseline number is for the incorrect behavior.

Test Plan:
- ninja -C build torch_cpu torch_python install
- python test/test_functionalization.py -k test_inplace_on_strided_input_preserves_stride_for_as_strided
- python test/test_functionalization.py -k test_copy_stride_mismatch
- python issue repro with torch.compile(backend="aot_eager", fullgraph=True): eager and compiled both return tensor([1, 3, 5])
- git diff --cached --check
- lintrunner -a (fails on unrelated existing clang-tidy diagnostics in aten/src/ATen/FunctionalTensorWrapper.cpp)

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the DrCI failures in functionalization, FX reinplace, and AOT dispatch tests. The failures were expected-graph mismatches from the new stride-preserving copy nodes; the targeted tests now pass locally, with the AOT test run under a local torchvision import guard due to this checkout's torchvision CUDA-version mismatch.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and resubmitted it to rerun CI. The previously reported functionalization, dynamic-shapes, and AOT targeted tests pass locally on the rebased checkout; full CI is pending.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the DrCI failures. The test_copy_ and AOT failures were missing expected graph updates from the new stride-preserving copy nodes, and the dynamic-shapes failure came from forcing a data-dependent guard on an unbacked symbolic storage offset. The helper now only acts on statically-known metadata mismatches so unbacked symbolic offsets do not force guards; the targeted tests pass locally.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the remaining functionalization CI failures and the dynamic-shapes review concern. The diagonal expected graphs now include the stride-preserving copy nodes, unknown symbolic stride equality now takes the preserve path without raw symbolic guards, and storage-offset handling remains conservative for unbacked symbolic offsets. Targeted functionalization, dynamic-shapes, FX reinplace, AOT, and issue repro tests pass locally; full CI is pending.

**Comment by jansel** [comment]
My agent says I updated the PR to address the remaining functionalization expecttest CI failures and the dynamic-shapes review feedback. The stale expected graphs now include the stride-preserving copy/copy_ nodes, the storage-offset comment documents the intentional symbolic-offset behavior, and a dynamic aot_eager regression covers the strided-input as_strided_copy case. The full test/test_functionalization.py file passes locally after rebuilding; lintrunner still reports only pre-existing clang-tidy diagnostics in FunctionalTensorWrapper.cpp.

**Comment by jansel** [comment]
@claude review these changes

File: torchgen/gen_functionalization_type.py [hunk 1]
```diff
@@ -591,7 +591,8 @@ def wrap_propagate_mutations_and_return(
         updates.append(
             f"""\
   auto {outer_arg}_inner = at::functionalization::impl::from_functional_tensor({outer_arg});
-  at::functionalization::impl::replace_({outer_arg}, {inner_ret});
+  auto {outer_arg}_updated = at::functionalization::impl::maybe_preserve_strides({outer_arg}_inner, {inner_ret});
```
Comment (aorenste) [inline torchgen/gen_functionalization_type.py]: [blocker] My agent says: this fires for view tensors too, where it is provably a no-op. `sync()` two lines down calls `regenerate_from_base()`, which recomputes `value_` from the view chain, and `apply_update()` feeds the value into a `*_scatter` that reads only its data -- so a view's layout never comes from `value_` in the first place. (The commented-out fast path in `commit_update()`, `FunctionalTensorWrapper.cpp:161-168`, was disabled for exactly this reason.)

Verified on `main` without this PR: `base[::2].add_()`, `base.diagonal().add_()`, `as_strided(b,[2],[2],1).add_()`, `split(b,2)[1].diagonal().add_()`, and `b.transpose(1,0).select(0,0).add_()` each followed by `as_strided_copy` already match eager. 15 of the 16 expecttests updated here are those view cases, so those copies are pure overhead -- and not cheap, since `copy` -> `clone_preserve_strides` clones the *whole* base storage. Replaying the updated `test_diagonal` graph shape on a 4096x4096 input: 10.6 ms -> 21.8 ms (2.06x). `aten.copy` on a 1-element view of a 1M-element storage allocates 4 MB.

Gate on the existing `isBaseTensor` (`FunctionalTensorWrapper.h:186`):

```python
  auto {outer_arg}_updated = at::functionalization::impl::maybe_preserve_strides({outer_arg}, {outer_arg}_inner, {inner_ret});
```

with both `maybe_preserve_strides` overloads taking the functional tensor(s) as a leading argument and returning `new_value` unchanged when `!isBaseTensor(...)`. That keeps the issue repro fixed and the false-aliasing graph at `test_aotdispatch.py:3136` as you have it, while reverting the other 15 expecttests to their pre-PR form.

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 2]
```diff
@@ -637,6 +638,72 @@ void replace_(const ITensorListRef functional_tensor, ITensorListRef other) {
...
+  if (!old_value.has_storage()) {
+    return new_value;
+  }
+  auto old_sizes = old_value.sym_sizes();
+  auto new_sizes = new_value.sym_sizes();
+  if (old_sizes.size() != new_sizes.size()) {
+    return new_value;
+  }
+  for (const auto dim : c10::irange(old_sizes.size())) {
+    if (TORCH_STATICALLY_KNOWN_TRUE(sym_ne(old_sizes[dim], new_sizes[dim]))) {
```
Comment (aorenste) [inline aten/src/ATen/FunctionalTensorWrapper.cpp]: [blocker] My agent says: opposite polarity from the stride check below -- this bails only when a size difference is *provable*, so an unknown-vs-known size falls through to the `copy`, which broadcasts silently (`aten.copy(zeros(3), ones(1))` returns `[1., 1., 1.]`, not a size-1 tensor) or hard-errors. Reachable via `out=` ops, which are allowed to resize the output (see the `replace_` comment at `:222`); for an unbacked size, `statically_known_true(0 != u0)` is `False`.

```cpp
    if (!TORCH_STATICALLY_KNOWN_TRUE(sym_eq(old_sizes[dim], new_sizes[dim]))) {
      return new_value;
    }
```

File: test/functorch/test_aotdispatch.py [hunk 3]
```diff
@@ -3232,12 +3234,21 @@ def inp_callable_overlap3(req_grad):
...
+            fw_graph2,
+            fw_graph3,
+            fw_graph4,
+            fw_graph5,
+            fw_graph6,
+        ):
+            self.assertTrue(
+                "def forward(self, arg0_1, arg1_1):" in str(non_overlap_graph.code)
+            )
+            self.assertFalse("as_strided_scatter" in str(non_overlap_graph.code))
```
Comment (aorenste) [inline test/functorch/test_aotdispatch.py]: [cleanup] My agent says: `assertTrue(x in y)` / `assertFalse(x in y)` print no diff on failure.

```python
            self.assertIn("def forward(self, arg0_1, arg1_1):", str(non_overlap_graph.code))
            self.assertNotIn("as_strided_scatter", str(non_overlap_graph.code))
```

Coverage here drops from "these six graphs are byte-identical" to "two args, no `as_strided_scatter`". If any subset of these inputs still ends up with the same layout, keeping equality within that subset would hold on to most of the original signal.

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] My agent says: only `pull`, `Lint`, and `docs-build` ran on this PR -- no `ciflow/inductor`, no `ciflow/trunk`. This changes the graph produced by *every* functionalized in-place mutation (16 expecttests updated here), and inductor's reinplacing pass plus the AOTI graph tests consume exactly those graphs. Please add `ciflow/inductor` and `ciflow/trunk` before merge.

[cleanup] My agent says: `topic: not user facing` will keep this out of the release notes, but it turns silently-wrong results into correct ones and adds a measurable slowdown for in-place mutations on non-contiguous tensors. Dropping that label seems right.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed all three review threads: view-only copies are skipped, unknown output sizes no longer enter an unsafe copy, and the AOT assertions retain exact equality where applicable. I also excluded metadata-mutating resize_as_ and set_ schemas while preserving generated out= layout semantics. The rebuilt functionalization suite (135 tests), targeted AOT/FX tests, and the prior issue repros pass. ciflow/inductor and ciflow/trunk are now running on the updated head.

**Comment by jansel** [comment]
My agent says BC Lint flagged the new torchgen preserve_strides parameter as required. I made it optional with a behavior-preserving default, reran lintrunner, and resubmitted the same one-commit stack; generated kernels and runtime behavior are unchanged.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
